File: regression_tester.rb
#!/usr/bin/env ruby
require 'fileutils'
print("Regression tester 0.0.1\n")
def path_local(from)
    
    return from #from.gsub('/','//')
    from.gsub('/','\\\\')
end
TESTS_DIR="./tests"
def perform_test(exepath,filepath,outname,args)
	output_path=path_local(TESTS_DIR+"/outputs/"+outname)
	exepath=path_local(exepath)
	output_path=path_local(output_path)
	filepath=path_local(filepath)
	joined_args = args.join(' ')
	printf("calling:" + "#{exepath} -a1 #{joined_args} -o#{output_path}.a1 #{filepath}\n")
	result = `#{exepath} -a1 -o#{output_path}.a1 #{filepath}`
	result = `#{exepath} -a2 #{joined_args} -o#{output_path}.a2 #{filepath}`
	result = `#{exepath} #{joined_args} -o#{output_path} #{filepath}`
	puts result
	p $?
end
`rm -rf #{TESTS_DIR}/outputs/*.*`
#exit(1)
Dir.open(TESTS_DIR+"/inputs").each() {|f|
	next if f=="." or f==".."
	perform_test(".//"+ARGV[0],TESTS_DIR+"/inputs/"+f,f,ARGV[1..-1])
}
Dir.open(TESTS_DIR+"/inputs").each() {|f|
	next if f=="." or f==".."
	FileUtils.mv(TESTS_DIR+"/inputs/"+f,TESTS_DIR+"/outputs/"+f) if f.end_with?(".b")
}
puts "**************************************\n"

File: tools/regression_tester/regression_tester.cpp
#include <QtCore/QCoreApplication>
#include <QProcess>
#include <QDir>
#include <QDirIterator>
#include <QDebug>
#include <QFileInfo>

static void perform_test(QString exepath,QString filepath, QStringList args) {
    const QString tgtpathbase("./tests/outputs/"+QFileInfo(filepath).completeBaseName());

    QProcess p;
    p.setProgram(exepath);
    QStringList assembly1_args = { QString("-a 1"), QString("-o" + tgtpathbase + ".a1"), filepath };
    p.setArguments(assembly1_args);
    p.start();
    if(!p.waitForFinished(30000)) {
        qCritical() << "Timeout while calling"<<p.program()<<"with"<<p.arguments();
    }

    QStringList assembly2_args = { QString("-a 2"), QString("-o" + tgtpathbase + ".a2"), filepath };
    p.setArguments(assembly2_args);
    p.start();
    if(!p.waitForFinished(30000)) {
        qCritical() << "Timeout while calling"<<p.program()<<"with"<<p.arguments();
    }

    QStringList decompile_args;
    decompile_args.append(args);
    decompile_args.append({ QString("-o" + tgtpathbase), filepath });
    p.setArguments(decompile_args);
    p.start();
    if(!p.waitForFinished(30000)) {
        qCritical() << "Timeout while calling"<<p.program()<<"with"<<p.arguments();
    }

    qDebug().noquote().nospace() << "Tests for "<<filepath<<" done.";
}
int main(int argc,char **argv) {
    QCoreApplication app(argc,argv);
    QByteArray z;
    QString TESTS_DIR="./tests";
    QStringList test_file_filter = {QString("*.exe"),QString("*.EXE")};
    QDirIterator input_iter(TESTS_DIR+"/inputs",test_file_filter,QDir::Filter::Files,QDirIterator::Subdirectories);
    QStringList orig_args=app.arguments();

    qDebug().noquote() <<"Regression tester 0.0.2";
    orig_args.pop_front(); // remove this exe name from args
    if(orig_args.size()<2) {
        return -1;
    }
    QString dcc_exe_path = orig_args.takeFirst();
    while(input_iter.hasNext()) {
        QString next_entry = input_iter.next();
        perform_test(dcc_exe_path,next_entry,orig_args);
    }
    qDebug().noquote() <<"**************************************";
}
